state-sync/src/coordinator.rs

```diff
@@ -1325,22 +1380,34 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
             );
         }
 
-        // verify the end-of-epoch LI for the following before passing it to execution:
-        // * verify end-of-epoch-li against local state
-        // * verify end-of-epoch-li's version corresponds to end-of-chunk version before
-        // passing it to execution
-        // * Executor expects that when it is passed an end-of-epoch LI, it is going to execute/commit
-        // transactions leading up to that end-of-epoch LI
-        // * verify end-of-epoch-li actually ends an epoch
-        let end_of_epoch_li = end_of_epoch_li
-            .map(|li| self.local_state.verify_ledger_info(&li).map(|_| li))
-            .transpose()?
-            .filter(|li| {
-                li.ledger_info().version() == known_version && li.ledger_info().ends_epoch()
-            });
+        // Verify the end_of_epoch_li against local state and ensure the version
+        // corresponds to the version at the end of the chunk.
+        // The executor expects that when it is passed an end_of_epoch_li, it is going
+        // to execute/commit transactions leading up to that li, so we
+        // also verify that the end_of_epoch_li actually ends the epoch.
+        let mut verified_end_of_epoch_li = None;
+        if let Some(end_of_epoch_li) = end_of_epoch_li {
+            self.local_state.verify_ledger_info(&end_of_epoch_li)?;
+
+            let ledger_info = end_of_epoch_li.ledger_info();
+            if !ledger_info.ends_epoch() {
+                return Err(Error::ProcessInvalidChunk(
+                    "Received waypoint ledger info with an end_of_epoch_li that does not end the epoch!".into(),
+                ));
+            }
+
+            // If we're now at the end of epoch version, use the ledger info to update the
+            // committed version in storage.
+            if ledger_info.version() == known_version {
+                verified_end_of_epoch_li = Some(end_of_epoch_li);
+            }
```

Comment:
you can use if / else to return back the Some or None.

Reply:
also I don't understand this test without looking deeper at how known_version was set, maybe we need to redefine this variable name and the function that you're calling?

Reply:
(I think this is okay, but it isn't super intuitive :P) you might want a doc block at the top of this function

Reply:
Yeah, this was deeply hidden in the existing code. I've updated the comment above the check and tried to call out that we might not be at the end_of_epoch_li, so we should only send it to storage (to be used to commit the state) if we're there. Does the comment make it clearer? Any suggestions, if not?

Reply:
yeah, the naming isn't intuitive, "verified" implies that it isn't verified until we do that last version check, what we really want is to call it (to_)commit_end_of_epoch_li or something like that.

Reply:
or eoeli_to_commit

Reply:
Done 😄